[incubator-kie-drools-5918] [new-parser] Annotations can appear in lh…

tkobayas opened the pull request:
…sOr and lhsAnd

**Issue**:
* https://github.com/apache/incubator-kie-drools/issues/5918

tkobayas commented on drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/antlr4/DRLVisitorImpl.java:
```diff
@@ -1032,7 +1038,13 @@ private BaseDescr createAndDescr(ParserRuleContext ctx) {
             // So, we need to flatten it so that AndDescr has A(), B() and C() as children.
             List<BaseDescr> flattenedDescrs = flattenAndDescr(descrList);
-            flattenedDescrs.forEach(andDescr::addDescr);
+            flattenedDescrs.forEach(descr -> {
+                if (descr instanceof AnnotationDescr annotationDescr) {
+                    andDescr.addAnnotation(annotationDescr);
+                } else {
+                    andDescr.addDescr(descr);
+                }
+            });
```
I know there is a duplication in this part and the above one (`visitLhsOr`). But `addDescr` method is not defined in a common parent class of `AndDescr` and `OrDescr`, so I cannot write a reusable clean private method. (Changing Descr class hierarchy sounds also too much).

tkobayas commented:
@yurloc @mariofusco @gitgabrio Please review, thanks!

yurloc commented on drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java:
```diff
@@ -3990,6 +3990,50 @@ void lhsPatternAnnotation() {
 
+    @Test
+    void andDescrAnnotation() {
+        ParserTestUtils.enableOldParser();
```
Committed unintentionally, I believe.

tkobayas replied:
Thank you! Fixed.

yurloc reviewed (changes requested):
To be compliant with the old parser, we also need to process annotations from each expression joined by infix OR/AND:
```
rule R
    when
       ( Double() 
         and @Annot1 String() 
         and @Annot2 Integer() ) 
    then
end
```

tkobayas commented:
> To be compliant with the old parser, we also need to process annotations from each expression joined by infix OR/AND:

Thank you! I added test cases and also fixed the implementation to collect the annotations.

yurloc reviewed (approved):
Perfect! Let's merge.
